File: Lecture02/Activities/app/src/main/java/com/codingblocks/activities/MainActivity.java
package com.codingblocks.activities;

import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;

public class MainActivity extends AppCompatActivity implements View.OnClickListener {
    EditText etVar1, etVar2;
    Button btnAdd, btnSub, btnMult, btnDiv;
    TextView tvResult;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        etVar1 = findViewById(R.id.etVar1);
        etVar2 = findViewById(R.id.etVar2);
        btnAdd = findViewById(R.id.btnAdd);
        btnSub = findViewById(R.id.btnSub);
        btnMult = findViewById(R.id.btnMult);
        btnDiv = findViewById(R.id.btnDiv);
        tvResult = findViewById(R.id.tvResult);

        btnAdd.setOnClickListener(this);
        btnSub.setOnClickListener(this);
        btnMult.setOnClickListener(this);
        btnDiv.setOnClickListener(this);


    }

    @Override
    public void onClick(View v) {
        try {
            int a = Integer.valueOf(etVar1.getText().toString());
            int b = Integer.valueOf(etVar2.getText().toString());
            int c = 0;
            switch (v.getId()) {
                case R.id.btnAdd:
                    c = a + b;
                    break;
                case R.id.btnSub:
                    c = a - b;
                    break;
                case R.id.btnMult:
                    c = a * b;
                    break;
                case R.id.btnDiv:
                    c = a / b;
                    break;
            }
            tvResult.setText(String.valueOf(c));
        } catch (NumberFormatException | ArithmeticException e) {
            Toast.makeText(this, "Wrong input or operation", Toast.LENGTH_SHORT).show();
        }
    }
}

File: Lecture17/Camera/app/src/main/java/com/codingblocks/camera/MainActivity.kt
package com.codingblocks.camera

import android.Manifest
import android.content.pm.PackageManager
import android.hardware.Camera
import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.os.Environment
import android.os.PersistableBundle
import android.os.SystemClock
import android.support.v4.app.ActivityCompat
import android.util.Log
import android.view.Surface
import android.view.SurfaceHolder
import android.widget.Toast
import kotlinx.android.synthetic.main.activity_main.*
import java.io.File

class MainActivity : AppCompatActivity() {
    val TAG = "CAM"
    lateinit var cam: Camera

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        ActivityCompat.requestPermissions(
                this,
                arrayOf(
                        Manifest.permission.CAMERA,
                        Manifest.permission.READ_EXTERNAL_STORAGE,
                        Manifest.permission.WRITE_EXTERNAL_STORAGE
                ),
                101
        )

        fun startPreview () {
            cam.setPreviewDisplay(surfaceView.holder)
            when (windowManager.defaultDisplay.rotation) {
                Surface.ROTATION_0 -> cam.setDisplayOrientation(90)
                Surface.ROTATION_90 -> cam.setDisplayOrientation(0)
                Surface.ROTATION_180 -> cam.setDisplayOrientation(270)
                Surface.ROTATION_270 -> cam.setDisplayOrientation(180)
            }
            cam.startPreview()
        }

        surfaceView.holder.addCallback(object: SurfaceHolder.Callback {
            override fun surfaceChanged(holder: SurfaceHolder?, format: Int, width: Int, height: Int) {
                holder?.surface?.let {
                    try {
                        cam.stopPreview()
                    } catch (e: Exception) {
                        Log.e(TAG, "No preview to stop", e)
                    }

                    try {
                        startPreview()
                    } catch (e: Exception) {
                        Log.e(TAG, "No preview to stop", e)
                    }
                }
            }

            override fun surfaceDestroyed(holder: SurfaceHolder?) {
            }

            override fun surfaceCreated(holder: SurfaceHolder?) {
            }

        })

        btnPreview.setOnClickListener {
            cam = Camera.open()
            Log.d(TAG, cam.toString())
            cam.parameters.pictureSize.let {
                Log.d(TAG, "${it.width}x${it.height}")
            }
            cam.parameters.previewSize.let {
                Log.d(TAG, "${it.width}x${it.height}")
            }
            startPreview()
        }

        btnCapture.setOnClickListener {
            cam.takePicture(
                    null,
                    null,
                    object: Camera.PictureCallback {
                        override fun onPictureTaken(data: ByteArray?, camera: Camera?) {
                            Log.d(TAG, "picsize = ${data?.size}")
                            if (data == null) return
                            val timeStamp = System.currentTimeMillis()
                            with(File(Environment.getExternalStorageDirectory(), "pic-$timeStamp.jpg")) {
                                writeBytes(data)
                            }
                            startPreview()

                        }

                    }
            )
        }


    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<out String>, grantResults: IntArray) {

        if (requestCode == 101) {
            if (grantResults.reduce{ acc, i -> acc+i } == 0) {

            } else {
                Toast.makeText(this, "CANNOT USE APP WITHOUT CAMERA PERM", Toast.LENGTH_SHORT)
                        .show()
                finish()
            }
        }

        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
    }
}
